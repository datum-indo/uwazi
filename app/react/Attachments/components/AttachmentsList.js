import PropTypes from 'prop-types';
import React, { Component } from 'react';
import { connect } from 'react-redux';

import { advancedSort } from 'app/utils/advancedSort';
import { t } from 'app/I18N';

import { NeedAuthorization } from 'app/Auth';
import Attachment from 'app/Attachments/components/Attachment';
import UploadAttachment from 'app/Attachments/components/UploadAttachment';
import UploadButton from 'app/Metadata/components/UploadButton';
import ViewDocButton from 'app/Library/components/ViewDocButton';
import Tip from '../../Layout/Tip';

export class AttachmentsList extends Component {
  getExtension(filename) {
    return filename.substr(filename.lastIndexOf('.') + 1);
  }

  arrangeFiles(files, isDocumentAttachments) {
    if (!files.length) {
      return files;
    }

    const firstFiles = [];
    if (isDocumentAttachments) {
      firstFiles.push(files.shift());
    }

    const sortedFiles = advancedSort(files, { property: 'originalname' });
    return firstFiles.concat(sortedFiles);
  }

  renderMainDocument(mainFile) {
    const { parentId, parentSharedId, readOnly, storeKey } = this.props;
    const forcedReadOnly = readOnly || Boolean(this.props.isTargetDoc);
    if (mainFile) {
      mainFile._id = parentId;
      return (
        <div>
          <h2>{t('System', 'Document')}</h2>
          <div className="attachments-list">
            <Attachment
              file={mainFile}
              parentId={parentId}
              readOnly={forcedReadOnly}
              storeKey={storeKey}
              parentSharedId={parentSharedId}
              isSourceDocument
              deleteMessage="Warning, Deleting the main file will also delete table of content and main files for the other languages of this entity"
            />
          </div>
          {this.props.entityView && mainFile && (
            <ViewDocButton
              file={mainFile}
              sharedId={parentSharedId}
              processed={this.props.processed}
              storeKey={storeKey}
            />
          )}
        </div>
      );
    }

    if (!forcedReadOnly) {
      return (
        <NeedAuthorization>
          <div className="attachment-buttons main-file">
            <h2>
              {t('System', 'Document')}
              <Tip>Main file: add a file as the main content</Tip>
            </h2>
            <UploadButton
              documentId={parentId}
              documentSharedId={parentSharedId}
              storeKey={storeKey}
            />
          </div>
        </NeedAuthorization>
      );
    }

    return null;
  }

  render() {
    const { parentId, parentSharedId, isDocumentAttachments, readOnly, storeKey } = this.props;
    const sortedFiles = this.arrangeFiles(this.props.files.toJS(), isDocumentAttachments);
    const forcedReadOnly = readOnly || Boolean(this.props.isTargetDoc);

    let uploadAttachmentButton = null;
    if (!this.props.isTargetDoc) {
      uploadAttachmentButton = (
        <NeedAuthorization roles={['admin', 'editor']}>
          <div className="attachment-add">
            <UploadAttachment entity={this.props.parentId} storeKey={storeKey} />
          </div>
        </NeedAuthorization>
      );
    }

    const mainFile = isDocumentAttachments ? sortedFiles[0] : null;
    const attachments = sortedFiles.filter((_f, index) => (mainFile && index !== 0) || !mainFile);
    return (
      <div className="attachments-list-parent">
        {this.renderMainDocument(mainFile)}
        <h2>{t('System', 'Attachments')}</h2>
        <div className="attachments-list">
          {attachments.map((file, index) => (
            <Attachment
              key={index}
              file={file}
              parentId={parentId}
              readOnly={forcedReadOnly}
              storeKey={storeKey}
              parentSharedId={parentSharedId}
              isSourceDocument={false}
            />
          ))}
        </div>
        {uploadAttachmentButton}
      </div>
    );
  }
}

AttachmentsList.propTypes = {
  files: PropTypes.object,
  parentId: PropTypes.string,
  model: PropTypes.string,
  parentSharedId: PropTypes.string,
  isDocumentAttachments: PropTypes.bool,
  readOnly: PropTypes.bool,
  entityView: PropTypes.bool,
  processed: PropTypes.bool,
  isTargetDoc: PropTypes.bool,
  deleteAttachment: PropTypes.func,
  loadForm: PropTypes.func,
  storeKey: PropTypes.string,
  user: PropTypes.object,
};

AttachmentsList.contextTypes = {
  confirm: PropTypes.func,
};

function mapStateToProps({ user }) {
  return {
    user,
    progress: null,
    model: 'documentViewer.sidepanel.attachment',
  };
}

export default connect(mapStateToProps)(AttachmentsList);
