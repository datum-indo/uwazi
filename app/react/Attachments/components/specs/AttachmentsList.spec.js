import React from 'react';
import { shallow } from 'enzyme';
import { fromJS as Immutable } from 'immutable';

import { AttachmentsList } from '../AttachmentsList';
import Attachment from '../Attachment';

describe('AttachmentsList', () => {
  let component;
  let props;
  let files;

  beforeEach(() => {
    files = Immutable([
      { originalname: 'Human name 1', filename: 'filename1.ext' },
      { originalname: 'A Human name 2', filename: 'filename2.ext' },
    ]);

    props = {
      files,
      parentId: 'parentId',
      parentSharedId: 'parentSharedId',
      isDocumentAttachments: false,
      readOnly: false,
      storeKey: 'storeKey',
      user: Immutable({}),
    };
  });

  const render = () => {
    component = shallow(<AttachmentsList {...props} />);
  };

  it('should render a sorted list of attachments (files)', () => {
    render();
    expect(component).toMatchSnapshot();
  });

  describe('When parent is Target Document', () => {
    beforeEach(() => {
      props.isTargetDoc = true;
      render();
    });

    it('should treat all Attachments as read only', () => {
      expect(component).toMatchSnapshot();
    });
  });

  describe('when files is empty', () => {
    it('should render nothing if user not logged in', () => {
      props.files = Immutable([]);
      render();
      expect(component).toMatchSnapshot();
    });

    it('should add button in Downloads section', () => {
      props.files = Immutable([]);
      props.user = Immutable({ _id: 'user' });
      render();
      expect(component).toMatchSnapshot();
    });
  });

  describe('When isDocumentAttachments', () => {
    beforeEach(() => {
      props.isDocumentAttachments = true;
      render();
    });

    it('should pass isSourceDocument true to the first attachment and the entity id', () => {
      expect(
        component
          .find(Attachment)
          .at(0)
          .props().isSourceDocument
      ).toBe(true);
      expect(
        component
          .find(Attachment)
          .at(0)
          .props().file._id
      ).toBe('parentId');
      expect(
        component
          .find(Attachment)
          .at(1)
          .props().isSourceDocument
      ).toBe(false);
    });
  });
});
