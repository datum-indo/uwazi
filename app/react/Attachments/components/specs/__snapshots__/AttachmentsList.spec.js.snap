// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AttachmentsList When parent is Target Document should treat all Attachments as read only 1`] = `
<div
  className="attachments-list-parent"
>
  <h2>
    Attachments
  </h2>
  <div
    className="attachments-list"
  >
    <Connect(Attachment)
      file={
        Object {
          "filename": "filename2.ext",
          "originalname": "A Human name 2",
        }
      }
      isSourceDocument={false}
      key="0"
      parentId="parentId"
      parentSharedId="parentSharedId"
      readOnly={true}
      storeKey="storeKey"
    />
    <Connect(Attachment)
      file={
        Object {
          "filename": "filename1.ext",
          "originalname": "Human name 1",
        }
      }
      isSourceDocument={false}
      key="1"
      parentId="parentId"
      parentSharedId="parentSharedId"
      readOnly={true}
      storeKey="storeKey"
    />
  </div>
</div>
`;

exports[`AttachmentsList should render a sorted list of attachments (files) 1`] = `
<div
  className="attachments-list-parent"
>
  <Connect(NeedAuthorization)>
    <div
      className="attachment-buttons main-file"
    >
      <h2>
        Document
        <Tip
          icon="question-circle"
        >
          Main file: add a file as the main content
        </Tip>
      </h2>
      <Connect(UploadButton)
        documentId="parentId"
        documentSharedId="parentSharedId"
        storeKey="storeKey"
      />
    </div>
  </Connect(NeedAuthorization)>
  <h2>
    Attachments
  </h2>
  <div
    className="attachments-list"
  >
    <Connect(Attachment)
      file={
        Object {
          "filename": "filename2.ext",
          "originalname": "A Human name 2",
        }
      }
      isSourceDocument={false}
      key="0"
      parentId="parentId"
      parentSharedId="parentSharedId"
      readOnly={false}
      storeKey="storeKey"
    />
    <Connect(Attachment)
      file={
        Object {
          "filename": "filename1.ext",
          "originalname": "Human name 1",
        }
      }
      isSourceDocument={false}
      key="1"
      parentId="parentId"
      parentSharedId="parentSharedId"
      readOnly={false}
      storeKey="storeKey"
    />
  </div>
  <Connect(NeedAuthorization)
    roles={
      Array [
        "admin",
        "editor",
      ]
    }
  >
    <div
      className="attachment-add"
    >
      <Connect(UploadAttachment)
        entity="parentId"
        storeKey="storeKey"
      />
    </div>
  </Connect(NeedAuthorization)>
</div>
`;

exports[`AttachmentsList when files is empty should add button in Downloads section 1`] = `
<div
  className="attachments-list-parent"
>
  <Connect(NeedAuthorization)>
    <div
      className="attachment-buttons main-file"
    >
      <h2>
        Document
        <Tip
          icon="question-circle"
        >
          Main file: add a file as the main content
        </Tip>
      </h2>
      <Connect(UploadButton)
        documentId="parentId"
        documentSharedId="parentSharedId"
        storeKey="storeKey"
      />
    </div>
  </Connect(NeedAuthorization)>
  <h2>
    Attachments
  </h2>
  <div
    className="attachments-list"
  />
  <Connect(NeedAuthorization)
    roles={
      Array [
        "admin",
        "editor",
      ]
    }
  >
    <div
      className="attachment-add"
    >
      <Connect(UploadAttachment)
        entity="parentId"
        storeKey="storeKey"
      />
    </div>
  </Connect(NeedAuthorization)>
</div>
`;

exports[`AttachmentsList when files is empty should render nothing if user not logged in 1`] = `
<div
  className="attachments-list-parent"
>
  <Connect(NeedAuthorization)>
    <div
      className="attachment-buttons main-file"
    >
      <h2>
        Document
        <Tip
          icon="question-circle"
        >
          Main file: add a file as the main content
        </Tip>
      </h2>
      <Connect(UploadButton)
        documentId="parentId"
        documentSharedId="parentSharedId"
        storeKey="storeKey"
      />
    </div>
  </Connect(NeedAuthorization)>
  <h2>
    Attachments
  </h2>
  <div
    className="attachments-list"
  />
  <Connect(NeedAuthorization)
    roles={
      Array [
        "admin",
        "editor",
      ]
    }
  >
    <div
      className="attachment-add"
    >
      <Connect(UploadAttachment)
        entity="parentId"
        storeKey="storeKey"
      />
    </div>
  </Connect(NeedAuthorization)>
</div>
`;
